test/unit/decoded_view.cpp

```diff
@@ -0,0 +1,508 @@
+        // empty()
+        {
+            decoded_view s;
+            BOOST_TEST(s.empty());
+
+            decoded_view s2(str);
+            BOOST_TEST_NOT(s2.empty());
+        }
+    }
+
+    void
+    testCopy()
+    {
+        // copy()
+        {
+            decoded_view s(str);
+            std::string out(s.size(), ' ');
+            s.copy(&out[0], s.size());
+            BOOST_TEST_EQ(s, dec_str);
+        }
+    }
+
+    void
+    testCompare()
+    {
+        // compare()
+        {
+            decoded_view s(str);
+            BOOST_TEST_EQ(s.compare(dec_str), 0);
+            BOOST_TEST_EQ(s.compare("a a"), 1);
+            BOOST_TEST_EQ(s.compare("a z"), -1);
+            std::string bs = "z";
+            BOOST_TEST_EQ(s.compare(bs), -1);
+        }
+
+        // operators
+        {
+            decoded_view s(str);
+
+            decoded_view s0(str);
+            decoded_view s1("a%20tri+test");
+            decoded_view s2("a%20vri+test");
+
+            string_view str0(dec_str);
+            string_view str1("a tri test");
+            string_view str2("a vri test");
+
+            std::string bstr0(dec_str);
+            std::string bstr1("a tri test");
+            std::string bstr2("a vri test");
+
+            BOOST_TEST(s == s0);
+            BOOST_TEST_NOT(s == s1);
+            BOOST_TEST(s == str0);
+            BOOST_TEST_NOT(s == str1);
+            BOOST_TEST(s == bstr0);
+            BOOST_TEST_NOT(s == bstr1);
+
+            BOOST_TEST(s != s2);
+            BOOST_TEST_NOT(s != s0);
+            BOOST_TEST(s != str2);
+            BOOST_TEST_NOT(s != str0);
+            BOOST_TEST(s != bstr2);
+            BOOST_TEST_NOT(s != bstr0);
+
+            BOOST_TEST(s < s2);
+            BOOST_TEST_NOT(s < s0);
+            BOOST_TEST(s < str2);
+            BOOST_TEST_NOT(s < str0);
+            BOOST_TEST(s < bstr2);
+            BOOST_TEST_NOT(s < bstr0);
+
+            BOOST_TEST(s <= s2);
+            BOOST_TEST(s <= s0);
+            BOOST_TEST(s <= str2);
+            BOOST_TEST(s <= str0);
+            BOOST_TEST(s <= bstr2);
+            BOOST_TEST(s <= bstr0);
+
+            BOOST_TEST(s > s1);
+            BOOST_TEST_NOT(s > s0);
+            BOOST_TEST(s > str1);
+            BOOST_TEST_NOT(s > str0);
+            BOOST_TEST(s > bstr1);
+            BOOST_TEST_NOT(s > bstr0);
+
+            BOOST_TEST(s >= s1);
+            BOOST_TEST(s >= s0);
+            BOOST_TEST(s >= str1);
+            BOOST_TEST(s >= str0);
+            BOOST_TEST(s >= bstr1);
+            BOOST_TEST(s >= bstr0);
+
+        }
+    }
+
+    void
+    testConversion()
+    {
+        // to_string()
+        {
+            decoded_view s(str);
+            std::string o = "to be recycled";
+            BOOST_TEST_EQ(s.to_string(o), dec_str);
+        }
+
+        // to_string()
+        {
+            decoded_view s(str);
+            BOOST_TEST_EQ(s.to_string(), dec_str);
+        }
+
+        // explicit operator basic_string()
```

Comment:
I don't think we should provide this function. The user should have to call `to_string()` explicitly. Otherwise it hides a potentially expensive operation. We should go to review without the function, and let the reviewers argue the case for including it if they think it is necessary.